File: aoc-py/solutions/day7.py
"""
Day 7: Bridge Repair

https://adventofcode.com/2024/day/7
"""
__all__ = ('Day7',)

from typing import ClassVar, Callable
from math import log10

from ..solution import Solution

class Day7(Solution):
    NAME: ClassVar[str] = 'Bridge Repair'

    def _solve_p1(self, target: int, terms: list[int], curr: int, progression: int) -> bool:
        if progression == len(terms):
            return target == curr

        term = terms[progression]

        return self._solve_p1(
            target,
            terms,
            curr + term,
            progression + 1,
        ) or self._solve_p1(
            target,
            terms,
            curr * term,
            progression + 1,
        )

    def _solve_p2(self, target: int, terms: list[int], curr: int, progression: int) -> bool:
        if progression == len(terms):
            return target == curr

        term = terms[progression]

        return self._solve_p2(
            target,
            terms,
            curr + term,
            progression + 1,
        ) or self._solve_p2(
            target,
            terms,
            curr * term,
            progression + 1,
        ) or self._solve_p2(
            target,
            terms,
            # `int(log10(term)) + 1` gives us the # of digits in the integer: `term`
            curr * 10 ** (int(log10(term)) + 1) + term,
            progression + 1,
        )

    def _solve(self, inp: str, solve_func: Callable[[int, list[int], int, int], bool]) -> int:
        total = 0

        for equation in inp.splitlines():
            target, terms = equation.split(':')

            first, *terms = [int(term) for term in terms.split()]
            target = int(target)

            if solve_func(target, terms, first, 0):
                total += target
        return total

    def part_one(self, inp: str) -> int:
        return self._solve(inp, self._solve_p1)

    def part_two(self, inp: str) -> int:
        return self._solve(inp, self._solve_p2)

    def run(self, inp: str) -> None:
        print('Part 1:', p1 := self.part_one(inp))
        print('Part 2:', p2 := self.part_two(inp))

        assert p1 == 3351424677624
        assert p2 == 204976636995111

File: src/bin/day7.rs
//! Day 7: Bridge Repair
//!
//! <https://adventofcode.com/2024/day/7>

use std::fmt::Display;
use aoc_2024::Solution;

pub struct Day7;

impl Day7 {
    fn solve_p1(target: usize, terms: &[usize], curr: usize, progression: usize) -> bool {
        if progression == terms.len() {
            return target == curr;
        }

        let term = terms[progression];

        Self::solve_p1(
            target,
            terms,
            curr + term,
            progression + 1
        ) || Self::solve_p1(
            target,
            terms,
            curr * term,
            progression + 1
        )
    }

    #[allow(
        clippy::cast_sign_loss,
        clippy::cast_precision_loss,
        clippy::cast_possible_truncation,
    )]
    fn solve_p2(target: usize, terms: &[usize], curr: usize, progression: usize) -> bool {
        if progression == terms.len() {
            return target == curr;
        }

        let term = terms[progression];

        Self::solve_p2(
            target,
            terms,
            curr + term,
            progression + 1,
        ) || Self::solve_p2(
            target,
            terms,
            curr * term,
            progression + 1,
        ) || Self::solve_p2(
            target,
            terms,
            curr * (10usize).pow((term as f64).log10() as u32 + 1) + term,
            progression + 1,
        )
    }
}

impl Solution for Day7 {
    const NAME: &'static str = "Bridge Repair";

    fn part_one<T: Display>(&self, inp: T) -> usize {
        inp.to_string()
            .lines()
            .filter_map(|equation| {
                let (target, terms) = equation.split_once(':').unwrap();

                let mut terms = terms
                    .split_whitespace()
                    .filter_map(|term| term.parse::<usize>().ok());
                let first = terms
                    .next()
                    .unwrap();
                let terms = terms
                    .collect::<Vec<_>>();
                let target = target
                    .parse()
                    .unwrap();

                Self::solve_p1(target, &terms, first, 0)
                    .then_some(target)
            })
            .sum()
    }

    fn part_two<T: Display>(&self, inp: T) -> usize {
        inp.to_string()
            .lines()
            .filter_map(|equation| {
                let (target, terms) = equation.split_once(':').unwrap();

                let mut terms = terms
                    .split_whitespace()
                    .filter_map(|term| term.parse::<usize>().ok());
                let first = terms
                    .next()
                    .unwrap();
                let terms = terms
                    .collect::<Vec<_>>();
                let target = target
                    .parse()
                    .unwrap();

                Self::solve_p2(target, &terms, first, 0)
                    .then_some(target)
            })
            .sum()
    }

    fn run(&self, inp: String) {
        let p1 = self.part_one(&inp);
        let p2 = self.part_two(&inp);

        println!("Part 1: {p1}");
        println!("Part 2: {p2}");

        assert_eq!(p1, 3_351_424_677_624);
        assert_eq!(p2, 204_976_636_995_111);
    }
}

fn main() {
    aoc_2024::run_day(7, &Day7);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() { main(); }
}